Add EIP712 ante

DracoLi opened the pull request:
Add support for EIP712 transactions

pirtleshell commented on app/ante/ante.go:
```diff
@@ -91,17 +94,27 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
 }
 
-func newCosmosAnteHandler(options HandlerOptions) sdk.AnteHandler {
+func newCosmosAnteHandler(options HandlerOptions, isEIP712 bool) sdk.AnteHandler {
```
is it possible to add `IsEIP712` as a field on the existing `HandlerOptions` struct?

DracoLi replied:
`HandlerOptions` is shared options for all ante handlers that gets passed during app setup so we can't put this option there. I can take a look at other refactors to clean up the params here.

DracoLi commented on app/ante/eip712_test.go:
```diff
@@ -0,0 +1,611 @@
+			},
+		},
+		{
+			name:           "fails when invalid fee payer is provided",
```
here are the fail cases

DracoLi replied:
it's pretty important to make sure this covers all the critical error cases. Thoughts on any additional error cases that should be added here? @rhuairahrighairidh @karzak @nddeluca @drklee3

drklee3 commented on x/evmutil/types/contract.go:
```diff
@@ -0,0 +1,48 @@
+	ERC20MintableBurnableJSON []byte
+
+	// ERC20MintableBurnableContract is the compiled erc20 contract
+	ERC20MintableBurnableContract evmtypes.CompiledContract
```
We could use a regular ERC20 abi here instead of the Mintable/Burnable one since the extra methods are no longer required

Here's the json for a plain ERC20 which can be a replacement for `ERC20MintableBurnable.json`

```json
{
  "abi": "[{\"inputs\":[{\"internalType\":\"string\",\"name\":\"name_\",\"type\":\"string\"},{\"internalType\":\"string\",\"name\":\"symbol_\",\"type\":\"string\"}],\"stateMutability\":\"nonpayable\",\"type\":\"constructor\"},{\"anonymous\":false,\"inputs\":[{\"indexed\":true,\"internalType\":\"address\",\"name\":\"owner\",\"type\":\"address\"},{\"indexed\":true,\"internalType\":\"address\",\"name\":\"spender\",\"type\":\"address\"},{\"indexed\":false,\"internalType\":\"uint256\",\"name\":\"value\",\"type\":\"uint256\"}],\"name\":\"Approval\",\"type\":\"event\"},{\"anonymous\":false,\"inputs\":[{\"indexed\":true,\"internalType\":\"address\",\"name\":\"from\",\"type\":\"address\"},{\"indexed\":true,\"internalType\":\"address\",\"name\":\"to\",\"type\":\"address\"},{\"indexed\":false,\"internalType\":\"uint256\",\"name\":\"value\",\"type\":\"uint256\"}],\"name\":\"Transfer\",\"type\":\"event\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"owner\",\"type\":\"address\"},{\"internalType\":\"address\",\"name\":\"spender\",\"type\":\"address\"}],\"name\":\"allowance\",\"outputs\":[{\"internalType\":\"uint256\",\"name\":\"\",\"type\":\"uint256\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"spender\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"amount\",\"type\":\"uint256\"}],\"name\":\"approve\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"\",\"type\":\"bool\"}],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"account\",\"type\":\"address\"}],\"name\":\"balanceOf\",\"outputs\":[{\"internalType\":\"uint256\",\"name\":\"\",\"type\":\"uint256\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"decimals\",\"outputs\":[{\"internalType\":\"uint8\",\"name\":\"\",\"type\":\"uint8\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"spender\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"subtractedValue\",\"type\":\"uint256\"}],\"name\":\"decreaseAllowance\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"\",\"type\":\"bool\"}],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"spender\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"addedValue\",\"type\":\"uint256\"}],\"name\":\"increaseAllowance\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"\",\"type\":\"bool\"}],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"name\",\"outputs\":[{\"internalType\":\"string\",\"name\":\"\",\"type\":\"string\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"symbol\",\"outputs\":[{\"internalType\":\"string\",\"name\":\"\",\"type\":\"string\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"totalSupply\",\"outputs\":[{\"internalType\":\"uint256\",\"name\":\"\",\"type\":\"uint256\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"to\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"amount\",\"type\":\"uint256\"}],\"name\":\"transfer\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"\",\"type\":\"bool\"}],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"from\",\"type\":\"address\"},{\"internalType\":\"address\",\"name\":\"to\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"amount\",\"type\":\"uint256\"}],\"name\":\"transferFrom\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"\",\"type\":\"bool\"}],\"stateMutability\":\"nonpayable\",\"type\":\"function\"}]",
  "bin": "60806040523480156200001157600080fd5b5060405162001706380380620017068339818101604052810190620000379190620002be565b81600390805190602001906200004f92919062000071565b5080600490805190602001906200006892919062000071565b505050620003a8565b8280546200007f9062000372565b90600052602060002090601f016020900481019282620000a35760008555620000ef565b82601f10620000be57805160ff1916838001178555620000ef565b82800160010185558215620000ef579182015b82811115620000ee578251825591602001919060010190620000d1565b5b509050620000fe919062000102565b5090565b5b808211156200011d57600081600090555060010162000103565b5090565b6000604051905090565b600080fd5b600080fd5b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6200018a826200013f565b810181811067ffffffffffffffff82111715620001ac57620001ab62000150565b5b80604052505050565b6000620001c162000121565b9050620001cf82826200017f565b919050565b600067ffffffffffffffff821115620001f257620001f162000150565b5b620001fd826200013f565b9050602081019050919050565b60005b838110156200022a5780820151818401526020810190506200020d565b838111156200023a576000848401525b50505050565b6000620002576200025184620001d4565b620001b5565b9050828152602081018484840111156200027657620002756200013a565b5b620002838482856200020a565b509392505050565b600082601f830112620002a357620002a262000135565b5b8151620002b584826020860162000240565b91505092915050565b60008060408385031215620002d857620002d76200012b565b5b600083015167ffffffffffffffff811115620002f957620002f862000130565b5b62000307858286016200028b565b925050602083015167ffffffffffffffff8111156200032b576200032a62000130565b5b62000339858286016200028b565b9150509250929050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052602260045260246000fd5b600060028204905060018216806200038b57607f821691505b60208210811415620003a257620003a162000343565b5b50919050565b61134e80620003b86000396000f3fe608060405234801561001057600080fd5b50600436106100a95760003560e01c80633950935111610071578063395093511461016857806370a082311461019857806395d89b41146101c8578063a457c2d7146101e6578063a9059cbb14610216578063dd62ed3e14610246576100a9565b806306fdde03146100ae578063095ea7b3146100cc57806318160ddd146100fc57806323b872dd1461011a578063313ce5671461014a575b600080fd5b6100b6610276565b6040516100c39190610c08565b60405180910390f35b6100e660048036038101906100e19190610cc3565b610308565b6040516100f39190610d1e565b60405180910390f35b61010461032b565b6040516101119190610d48565b60405180910390f35b610134600480360381019061012f9190610d63565b610335565b6040516101419190610d1e565b60405180910390f35b610152610364565b60405161015f9190610dd2565b60405180910390f35b610182600480360381019061017d9190610cc3565b61036d565b60405161018f9190610d1e565b60405180910390f35b6101b260048036038101906101ad9190610ded565b610417565b6040516101bf9190610d48565b60405180910390f35b6101d061045f565b6040516101dd9190610c08565b60405180910390f35b61020060048036038101906101fb9190610cc3565b6104f1565b60405161020d9190610d1e565b60405180910390f35b610230600480360381019061022b9190610cc3565b6105db565b60405161023d9190610d1e565b60405180910390f35b610260600480360381019061025b9190610e1a565b6105fe565b60405161026d9190610d48565b60405180910390f35b60606003805461028590610e89565b80601f01602080910402602001604051908101604052809291908181526020018280546102b190610e89565b80156102fe5780601f106102d3576101008083540402835291602001916102fe565b820191906000526020600020905b8154815290600101906020018083116102e157829003601f168201915b5050505050905090565b600080610313610685565b905061032081858561068d565b600191505092915050565b6000600254905090565b600080610340610685565b905061034d858285610858565b6103588585856108e4565b60019150509392505050565b60006012905090565b600080610378610685565b905061040c818585600160008673ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060008973ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001908152602001600020546104079190610eea565b61068d565b600191505092915050565b60008060008373ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001908152602001600020549050919050565b60606004805461046e90610e89565b80601f016020809104026020016040519081016040528092919081815260200182805461049a90610e89565b80156104e75780601f106104bc576101008083540402835291602001916104e7565b820191906000526020600020905b8154815290600101906020018083116104ca57829003601f168201915b5050505050905090565b6000806104fc610685565b90506000600160008373ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060008673ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001908152602001600020549050838110156105c2576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016105b990610fb2565b60405180910390fd5b6105cf828686840361068d565b60019250505092915050565b6000806105e6610685565b90506105f38185856108e4565b600191505092915050565b6000600160008473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060008373ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002054905092915050565b600033905090565b600073ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff1614156106fd576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016106f490611044565b60405180910390fd5b600073ffffffffffffffffffffffffffffffffffffffff168273ffffffffffffffffffffffffffffffffffffffff16141561076d576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401610764906110d6565b60405180910390fd5b80600160008573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060008473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001908152602001600020819055508173ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff167f8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b9258360405161084b9190610d48565b60405180910390a3505050565b600061086484846105fe565b90507fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff81146108de57818110156108d0576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016108c790611142565b60405180910390fd5b6108dd848484840361068d565b5b50505050565b600073ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff161415610954576040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161094b906111d4565b60405180910390fd5b600073ffffffffffffffffffffffffffffffffffffffff168273ffffffffffffffffffffffffffffffffffffffff1614156109c4576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016109bb90611266565b60405180910390fd5b6109cf838383610b65565b60008060008573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002054905081811015610a55576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401610a4c906112f8565b60405180910390fd5b8181036000808673ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002081905550816000808573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020016000206000828254610ae89190610eea565b925050819055508273ffffffffffffffffffffffffffffffffffffffff168473ffffffffffffffffffffffffffffffffffffffff167fddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef84604051610b4c9190610d48565b60405180910390a3610b5f848484610b6a565b50505050565b505050565b505050565b600081519050919050565b600082825260208201905092915050565b60005b83811015610ba9578082015181840152602081019050610b8e565b83811115610bb8576000848401525b50505050565b6000601f19601f8301169050919050565b6000610bda82610b6f565b610be48185610b7a565b9350610bf4818560208601610b8b565b610bfd81610bbe565b840191505092915050565b60006020820190508181036000830152610c228184610bcf565b905092915050565b600080fd5b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000610c5a82610c2f565b9050919050565b610c6a81610c4f565b8114610c7557600080fd5b50565b600081359050610c8781610c61565b92915050565b6000819050919050565b610ca081610c8d565b8114610cab57600080fd5b50565b600081359050610cbd81610c97565b92915050565b60008060408385031215610cda57610cd9610c2a565b5b6000610ce885828601610c78565b9250506020610cf985828601610cae565b9150509250929050565b60008115159050919050565b610d1881610d03565b82525050565b6000602082019050610d336000830184610d0f565b92915050565b610d4281610c8d565b82525050565b6000602082019050610d5d6000830184610d39565b92915050565b600080600060608486031215610d7c57610d7b610c2a565b5b6000610d8a86828701610c78565b9350506020610d9b86828701610c78565b9250506040610dac86828701610cae565b9150509250925092565b600060ff82169050919050565b610dcc81610db6565b82525050565b6000602082019050610de76000830184610dc3565b92915050565b600060208284031215610e0357610e02610c2a565b5b6000610e1184828501610c78565b91505092915050565b60008060408385031215610e3157610e30610c2a565b5b6000610e3f85828601610c78565b9250506020610e5085828601610c78565b9150509250929050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052602260045260246000fd5b60006002820490506001821680610ea157607f821691505b60208210811415610eb557610eb4610e5a565b5b50919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b6000610ef582610c8d565b9150610f0083610c8d565b9250827fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff03821115610f3557610f34610ebb565b5b828201905092915050565b7f45524332303a2064656372656173656420616c6c6f77616e63652062656c6f7760008201527f207a65726f000000000000000000000000000000000000000000000000000000602082015250565b6000610f9c602583610b7a565b9150610fa782610f40565b604082019050919050565b60006020820190508181036000830152610fcb81610f8f565b9050919050565b7f45524332303a20617070726f76652066726f6d20746865207a65726f2061646460008201527f7265737300000000000000000000000000000000000000000000000000000000602082015250565b600061102e602483610b7a565b915061103982610fd2565b604082019050919050565b6000602082019050818103600083015261105d81611021565b9050919050565b7f45524332303a20617070726f766520746f20746865207a65726f20616464726560008201527f7373000000000000000000000000000000000000000000000000000000000000602082015250565b60006110c0602283610b7a565b91506110cb82611064565b604082019050919050565b600060208201905081810360008301526110ef816110b3565b9050919050565b7f45524332303a20696e73756666696369656e7420616c6c6f77616e6365000000600082015250565b600061112c601d83610b7a565b9150611137826110f6565b602082019050919050565b6000602082019050818103600083015261115b8161111f565b9050919050565b7f45524332303a207472616e736665722066726f6d20746865207a65726f20616460008201527f6472657373000000000000000000000000000000000000000000000000000000602082015250565b60006111be602583610b7a565b91506111c982611162565b604082019050919050565b600060208201905081810360008301526111ed816111b1565b9050919050565b7f45524332303a207472616e7366657220746f20746865207a65726f206164647260008201527f6573730000000000000000000000000000000000000000000000000000000000602082015250565b6000611250602383610b7a565b915061125b826111f4565b604082019050919050565b6000602082019050818103600083015261127f81611243565b9050919050565b7f45524332303a207472616e7366657220616d6f756e742065786365656473206260008201527f616c616e63650000000000000000000000000000000000000000000000000000602082015250565b60006112e2602683610b7a565b91506112ed82611286565b604082019050919050565b60006020820190508181036000830152611311816112d5565b905091905056fea2646970667358221220256a622ed3bd1824651ebff077e536618fcc6063fb43c7252b78a1264311871464736f6c63430008090033"
}
```

Which was made with these commands in kava-bridge:

```bash
cd kava-bridge/contract/
npx hardhat compile 
jq '.abi = (.abi | tostring) | {abi, bin: .bytecode[2:] }' < artifacts/@openzeppelin/contracts/token/ERC20/ERC20.sol/ERC20.json > ethermint_json/ERC20.json
```

DracoLi replied:
I believe a lot of tests use mint to fund accounts. Since this contract is only used in tests, I think we can keep to avoid redoing a lot of tests that mints?

DracoLi replied:
synced on this, will keep it since tests uses mint

DracoLi commented on app/app.go:
```diff
@@ -392,6 +392,7 @@ func NewApp(
 	feemarketSubspace := app.paramsKeeper.Subspace(feemarkettypes.ModuleName)
 	evmSubspace := app.paramsKeeper.Subspace(evmtypes.ModuleName)
 	bridgeSubspace := app.paramsKeeper.Subspace(bridgetypes.ModuleName)
```
I think so. i can split that out into a separate PR since this one is quite large

rhuairahrighairidh commented on x/evmutil/keeper/evm.go:
```diff
@@ -0,0 +1,130 @@
+		big.NewInt(0), // gasPrice
+		data,
+		ethtypes.AccessList{}, // AccessList
+		true,                  // checkNonce
```
Do you know what the last arg is used for? It's called isFake which seems a bit odd

rhuairahrighairidh commented on x/evmutil/types/conversion_pair.go:
```diff
@@ -0,0 +1,89 @@
+
+// Validate returns an error if the ConversionPair is invalid.
+func (pair ConversionPair) Validate() error {
+	if pair.Denom == "" {
+		return errors.New("denom cannot be empty")
+	}
```
nit: `sdk.ValidateDenom` is used to validate denoms in other modules. I'd probably use it here in as we intend to send these coins to other modules.

rhuairahrighairidh commented on app/ante/ante.go:
```diff
@@ -91,17 +106,27 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
-		authante.NewRejectExtensionOptionsDecorator(),
 	)
+
+	if !options.isEIP712 {
+		decorators = append(decorators, authante.NewRejectExtensionOptionsDecorator())
+	}
```
nit: looks like it's possible to add both a `Web3Tx` and `EthereumTx` option in a valid tx. I'd probably limit the number of extensions to just one to remove the chance for confusion. Maybe up on lines 72-73

rhuairahrighairidh commented on x/evmutil/genesis.go:
```diff
@@ -11,6 +11,8 @@ import (
 
 // InitGenesis initializes the store state from a genesis state.
 func InitGenesis(ctx sdk.Context, keeper keeper.Keeper, gs *types.GenesisState) {
+	keeper.SetParams(ctx, gs.Params)
+
```
Any reason for setting params before genesis validation?

DracoLi replied:
yeah no reason, i will move this down

rhuairahrighairidh commented on x/evmutil/module.go:
```diff
@@ -104,7 +116,7 @@ func (am AppModule) Name() string {
 func (am AppModule) Route() sdk.Route { return sdk.Route{} }
 
 // QuerierRoute returns evmutil module's query routing key.
-func (AppModule) QuerierRoute() string { return types.ModuleName }
+func (AppModule) QuerierRoute() string { return types.RouterKey }
```
nit: `QuerierRoute` isn't used for modules only using grpc queriers. We could delete the `RouterKey` const and return "" here as [newer sdk modules do](https://github.com/cosmos/cosmos-sdk/blob/35ae2c4c72d4aeb33447d5a7af23ca47f786606e/x/authz/module/module.go#L129-L130)

rhuairahrighairidh commented on proto/kava/evmutil/v1beta1/tx.proto:
```diff
@@ -0,0 +1,52 @@
+  // EVM 0x hex address initiating the conversion.
+  string initiator = 1;
+  // Kava bech32 address that will receive the converted sdk.Coin.
+  string receiver = 2;
```
nit: in other modules we normally add `(cosmos_proto.scalar) = "cosmos.AddressString"` to sdk address strings. I don't think it does anything yet but probably worth keeping consistency.

rhuairahrighairidh commented on x/evmutil/types/msg.go:
```diff
@@ -0,0 +1,146 @@
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "receiver is not a valid bech32 address")
+	}
+
+	if msg.Amount.LTE(sdk.ZeroInt()) {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "amount cannot be zero or less")
+	}
```
nit/gotcha: `sdk.Int` contain a pointer to a `big.Int` which if it's nil will cause this line to panic. Not a big deal but can be avoided by checking `sdk.Int`'s `IsNil()` method.

rhuairahrighairidh commented on x/evmutil/types/address.go:
```diff
@@ -0,0 +1,61 @@
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// ExternalEVMAddress is a type alias of common.Address to represent an address
+// on an external EVM, e.g. Ethereum. This is used to make external / internal
+// addresses type safe and un-assignable to each other. This also makes it more
+// clear which address belongs where.
+type ExternalEVMAddress struct {
+	common.Address
+}
```
nit: Don't think we need this type anymore? so could be removed

rhuairahrighairidh commented on x/evmutil/keeper/erc20.go:
```diff
@@ -0,0 +1,144 @@
+// minting permissions for the module account.
+// Derived from tharsis/evmos
+// https://github.com/tharsis/evmos/blob/ee54f496551df937915ff6f74a94732a35abc505/x/erc20/keeper/evm.go
+func (k Keeper) DeployMintableERC20Contract(
+	ctx sdk.Context,
+	name string,
+	symbol string,
+	decimals uint8,
+) (types.InternalEVMAddress, error) {
```
nit: If this method is only used in tests for now I'd maybe make it a bit more obvious (like put "Test" in the name or move it off the main Keeper). Want to avoid lower quality test code accidentaly making it into production in a future PR. Somewhat moot as this code is already on mainnet, but we might want to review the contract code before using it in this module.

rhuairahrighairidh commented on x/evmutil/keeper/conversion.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	// lock erc20 tokens
+	if err := k.LockERC20Tokens(ctx, pair, amount.BigInt(), initiator); err != nil {
+		return err
+	}
```
ran into an edge case where if the contract for a pair in the params doesn't exist, the evm call will not return error, allowing the user to mint infinite coins.
This could happen if we mess up a gov proposal, or create one before the contract is deployed, or if an existing contract is suicided. I think we could check to see if the contract address points to a contract account to prevent this from happening.

rhuairahrighairidh reviewed (changes requested):
found an edge case thats worth preventing, but otherwise just nits

DracoLi commented:
@rhuairahrighairidh addressed feedback https://github.com/Kava-Labs/kava/pull/1267/files/ae7cd895aa1c0d342bd9df126af70d45aa0154a2..9b2dd56551747ff614b6775c6f332c0de0ecd059

rhuairahrighairidh commented on x/evmutil/types/msg.go:
```diff
@@ -123,7 +123,7 @@ func (msg MsgConvertERC20ToCoin) ValidateBasic() error {
 	}
 
-	if msg.Amount.LTE(sdk.ZeroInt()) {
+	if !msg.Amount.IsNil() && msg.Amount.LTE(sdk.ZeroInt()) {
```
```suggestion
	if msg.Amount.IsNil() || msg.Amount.LTE(sdk.ZeroInt()) {
```
looks like this lets a nil value through. I think an OR exits early if it finds a `true` so should never call `.LTE(...)` on a nil value.

rhuairahrighairidh reviewed (approved):
Looks good, we'll need to update our ethermint fork to include the fixes recently included in kava v0.18 but that could be done as a separate PR to help unblock webapp progress.
